File: LintCode-Python/addTwoNumbers.py
# You have two numbers represented by a linked list, where each node contains a
# single digit. The digits are stored in reverse order, such that the 1's digit
# is at the head of the list. Write a function that adds the two numbers and
#  returns the sum as a linked list.

# Have you met this question in a real interview? Yes
# Example
# Given 7->1->6 + 5->9->2. That is, 617 + 295.

# Return 2->1->9. That is 912.

# Given 3->1->5 and 5->9->2, return 8->0->8.


from ListNode import ListNode

def addTwoNumbers(l1, l2):
    if l1 is None and l2 is None:
        return None
    dummy = ListNode(0)
    point  = dummy
    carry = 0
    while l1 is not None and l2 is not None:
        sum = l1.val + l2.val + carry
        carry = sum / 10
        point.next = ListNode(sum % 10)
        l1 = l1.next
        l2 = l2.next
        point = point.next
    while l1 is not None:
        sum = l1.next + carry
        carry = sum / 10
        point.next = ListNode(sum % 10)
        l1 = l1.next
        point = point.next
    while l2 is not None:
        sum = l2.val + carry
        carry = sum / 10
        point.next = ListNode(sum % 10)
        l2 = l2.next
        point = point.next
    if carry != 0:
        point.next = ListNode(carry)
    return dummy.next




l1 = ListNode.arrayToList([7, 1, 6])
l2 = ListNode.arrayToList([5, 9, 2])
ListNode.printList(addTwoNumbers(l1, l2))

File: LeetCode_Java/src/AddTwoNumbers.java
/**
 * You are given two linked lists representing two non-negative numbers. The digits are stored in 
 * reverse order and each of their nodes contain a single digit. Add the two numbers and return it 
 * as a linked list.
 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 * Output: 7 -> 0 -> 8
 * 
 * Tags: LinkedList, Math
 * @author wendi
 *
 */
public class AddTwoNumbers {

	/**
	 * Brute force
	 * @param ListNode l1, ListNode l2
	 * @return ListNode
	 * Time: O(n1 + n2)
	 * Space: O(1)
	 */
	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		if (l1 == null) return l2;
		if (l2 == null) return l1;
		ListNode dummy = new ListNode(0);
		ListNode node = dummy;
		int carry = 0;
		while (l1 != null || l2 != null || carry != 0) {
			int val1 = l1 == null ? 0 : l1.val;
			int val2 = l2 == null ? 0 : l2.val;
			int sum = val1 + val2 + carry;
			node.next = new ListNode(sum % 10);
			carry = sum / 10;
			if (l1 != null) l1 = l1.next;
			if (l2 != null) l2 = l2.next;
			node = node.next;
		}
		return dummy.next;
	}
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		AddTwoNumbers result = new AddTwoNumbers();
		ListNode l1 = ListNode.generateLinkedList(new int[] {5, 4, 6});
		ListNode l2 = ListNode.generateLinkedList(new int[] {5, 2, 4, 9});
		ListNode.printLinkedList(l1);
		ListNode.printLinkedList(l2);
		ListNode.printLinkedList(result.addTwoNumbers(l1, l2));
	}

}
